integration/shareddata/iterator.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shareddata
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"hash"
+	"sync"
+
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/resource"
+)
+
+// newValuesIterator creates iterator that iterates through bson.D documents generated
+// by generator function.
+// Generator should return next deterministic bson.D document on every execution.
+// To stop iterator, generator must return nil.
+func newValuesIterator(generator func() bson.D) *valuesIterator {
+	iter := &valuesIterator{
+		token:     resource.NewToken(),
+		generator: generator,
+		hash:      sha256.New(),
+	}
+
+	resource.Track(iter, iter.token)
+
+	return iter
+}
+
+// valuesIterator iterates through bson.D documents created by generator function.
+// It also calculates the checksum of all documents on-fly.
+type valuesIterator struct {
+	generator func() bson.D
+	token     *resource.Token
+	hash      hash.Hash
+
+	// m Mutex protects generator function, and hash from parallel calls.
+	// It's not placed on top of struct fields because of field alignment.
+	m sync.Mutex
+}
+
+// Next implements iterator.Interface.
+func (iter *valuesIterator) Next() (struct{}, bson.D, error) {
+	iter.m.Lock()
+	defer iter.m.Unlock()
+
+	var unused struct{}
+
+	if iter.generator == nil {
+		return unused, nil, iterator.ErrIteratorDone
+	}
+
+	doc := iter.generator()
+	if doc == nil {
+		return unused, nil, iterator.ErrIteratorDone
+	}
+
+	rawDoc := []byte(fmt.Sprintf("%x", doc))
+
+	// append literal document to previous checksum.
+	//
+	// We don't just write all the documents and Sum at the end, to not
+	// store all documents in memory.
+	currSum := iter.hash.Sum(rawDoc)
+
+	// remove old checksum
+	iter.hash.Reset()
+
+	// write literal document with previous checksum to calculate new checksum
+	if _, err := iter.hash.Write(currSum); err != nil {
+		return unused, doc, err
+	}
```

Comment:
That's a very strange way to hash documents _if_ it works. Are you building a blockchain there? :)

All we need to do here is to call `iter.hash.Write(rawDoc)`

Reply:
I did it in a "blockchain" fashion, as it requires constant amount of memory, regardless from document amount. Writing all documents to the hash will result in storing all of them in memory, and as we use iterator here to save some space, I thought that doing that here is also reasonable.

Reply:
> Writing all documents to the hash will result in storing all of them in memory

Why?.. I'm not sure why you think that, but that's how hashing is typically used: https://go.dev/play/p/5HzWihaoJUP `Write` can be (and normally should be) called multiple times and `Sum` accepts a slice only as performance optimizations in some use cases (to avoid allocation).